event/parsing/totalBootTimeParser.go

```diff
@@ -0,0 +1,170 @@
+	if !destinationRegex.MatchString(msg.Destination) || !manageableRegex.MatchString(msg.Destination) {
+		level.Debug(b.Logger).Log(xlog.MessageKey(), "event is not an fully-manageable event")
+		return -1, nil
+	}
+
+	// Get boot time and device id from message.
+	bootTimeInt, deviceID, err := getWRPInfo(destinationRegex, msg)
+	if err != nil {
+		level.Error(b.Logger).Log(xlog.ErrorKey(), err)
+		return -1, err
+	}
+
+	// Get events from codex pertaining to this device id.
+	events := b.Client.GetEvents(deviceID)
+	latestPreviousEvent := Event{}
+
+	// Go through events to find reboot-pending event with the boot-time of the previous session
+	for _, event := range events {
+		if latestPreviousEvent, err = checkLatestPreviousEvent(event, latestPreviousEvent, bootTimeInt, rebootRegex); err != nil {
+			level.Error(b.Logger).Log(xlog.ErrorKey(), err)
+			if errors.Is(err, errNewerBootTime) {
+				// Something is wrong with this event's boot time, we shouldn't continue.
+				b.Measures.UnparsableEventsCount.With(ParserLabel, b.Label, ReasonLabel, eventBootTimeErr).Add(1.0)
+				return -1, err
+			}
+		}
+	}
+
+	if valid, err := isEventValid(latestPreviousEvent, rebootRegex, time.Now); !valid {
+		level.Error(b.Logger).Log(xlog.ErrorKey(), err)
+		return -1, err
+	}
+
+	// boot time calculation
+	// Event birthdate is saved in unix nanoseconds, so we must first convert it to a unix time using nanoseconds.
+	restartTime := wrpWithTime.Beginning.Sub(time.Unix(0, latestPreviousEvent.BirthDate)).Seconds()
+
+	if restartTime <= 0 {
+		err = errInvalidRestartTime
+		level.Error(b.Logger).Log(xlog.ErrorKey(), err, "Restart time", restartTime)
+		return -1, err
+	}
+
+	return restartTime, nil
+}
+
+// sees if this event is part of the most recent previous session
+func checkLatestPreviousEvent(e Event, previousEventTracked Event, latestBootTime int64, eventType *regexp.Regexp) (Event, error) {
+	eventBootTimeInt, err := GetEventBootTime(e)
+	previousEventBootTime, _ := GetEventBootTime(previousEventTracked)
+
+	if err != nil {
+		return previousEventTracked, err
+	}
+
+	if eventBootTimeInt > latestBootTime {
+		return Event{}, errNewerBootTime
+	}
+
+	// If this event has a boot time greater than what we've seen so far
+	// but still less than the current boot-time, it means that this event
+	// is part of a more recent previous cycle.
+	if eventBootTimeInt > previousEventBootTime && eventBootTimeInt < latestBootTime {
+		return e, nil
+	}
+
+	// If the event is of the type we are looking for and has the same boot-time as the
+	// newest previous boot time, return this event.
+	if eventBootTimeInt == previousEventBootTime && eventType.MatchString(e.Dest) {
+		// If the previously tracked event doesn't match the event we're looking for,
+		// return the current event. If both events match the type of event we are looking for,
+		// compare the birthdates and return the one with the more recent birthdate.
```

Comment:
I'm not sure this is the functionality we want when there are two `reboot-pending` events with the same boot time.  If somehow a customer tells a device to reboot twice, the first event is likely when the device actually started rebooting, right? Or finding two events could be considered an error of sorts and we could treat this differently (either in a failed metric or just labelling it in the normal metric).

This is a @schmidtw decision.

Reply:
If we find two `reboot-pending` events the older event is where we should start the measurement from.